### DISCUSSION TITLE
Two ore more indicators on one page

### DISCUSSION THREAD

**Discussion by christoph-LSN** [open-sdg/open-sdg#2013] [Ideas]
A new group of customers (policy decision makers) are interested in my migration report https://christoph-lsn.github.io/MT_Site/ . 
They want to have a quick overview of two or more (max 10) indicators on one page. Only graphs, tables or maps should be shown. The indicators schould be listed among themselves

like

- Goal 1
- Indicator 1-1-1
- Indicator 1-3-1
- Goal 2
- Indicator 2-2-1
- Indicator 2-2-4

I have tried to get the relevant html code from the indicator page with this code: 

```
{% for goal in page.goals %}
  {% if goal.number == '2' %}
    <h2>{{ goal.number }} {{ goal.name }}</h2>
    {% for indicator in page.indicators %}
      {% if indicator.number == '2.2.1' or indicator.number == '2.2.4' %}
        <table width="100%">
          <col style="width:90%">
          <col style="width:10%">
          <tr>
            <td><a href="{{ indicator.url }}">{{ indicator.number }} {{ indicator.name }}</a></td>
          </tr>
          <tr>
            <td>
              <div id="chartview">
                {% assign indicator_page_url = indicator.url | append: '/' %}
                {% assign indicator_page_content = indicator_page_url | get %}
                {% assign diagram_html = indicator_page_content | split: '<div id="chartview">' | last | split: '</div>' | first | prepend: '<div id="chartview">' | append: '</div>' %}
                {{ diagram_html }}
              </div>
            </td>
          </tr>
        </table>
      {% endif %}
    {% endfor %}
  {% endif %}
{% endfor %}

```

But it doesn't work as you can see here: https://christoph-lsn.github.io/MT_Site/decision_maker/

Two questions: 
@brockfanning
1) Is it right, that the indicator page is a static page containing graph and table so that I can grab it from the indicator page?  

2)  Additionally they need to print the site. We had a discussion about printing  here: #1436 almost two years ago. @MoBosse do you have any progress in your printing issues?

**Comment by christoph-LSN** [comment]
Small update:

I included some files... which is workung more or less.... somthing is displayed... I think I haven't checked how to connect the site to the data

```
{% for goal in page.goals %}
  {% if goal.number == '2' %}
    <h2>{{ goal.number }} {{ goal.name }}</h2>
    {% for indicator in page.indicators %}
      {% if indicator.number == '2.2.1' or indicator.number == '2.2.4' %}
        <table width="100%">
          <col style="width:90%">
          <col style="width:10%">
          <tr>
            <td><a href="{{ indicator.url }}">{{ indicator.number }} {{ indicator.name }}</a></td>
          </tr>
          <tr>
            <td>





{% include components/indicator/fields-template.html %}
{% include components/indicator/units-template.html %}
{% include components/indicator/series-template.html %}
{% include components/indicator/data-main.html %}
{% include components/indicator/data-tabs.html %}
{% include components/indicator/data-panes.html %}
{% include components/indicator/data-sidebar.html %}
{% include components/indicator/headline.html %}
{% include components/indicator/map.html %}
{% include components/indicator/table.html %}
{% include multilingual-js.html key="indicator" %}
{% include multilingual-js.html key="data" %}


<div id="main-content" class="container goal-{{ page.goal.number }}" role="main">
  {% include components/indicator/data-notice.html %}
  {% include components/indicator/indicator-available.html %}
  {% include components/indicator/indicator-progress.html %}
  {% include components/indicator/indicator-content.html %}
  {% include components/indicator/indicator-main.html %}
  
</div>


              </div>              
            </td>
          </tr>
        </table>
      {% endif %}
    {% endfor %}
  {% endif %}
{% endfor %}
```

![image](https://github.com/open-sdg/open-sdg/assets/94443931/5aeba083-b032-4a3f-9ef8-944c4fda94c2)

**Comment by christoph-LSN** [comment]
Another Update: I tried to grab the code from the indicator page with this code:

```
<script src="https://code.jquery.com/jquery-3.6.0.min.js"></script>
  <script>
    $(document).ready(function() {
      $.ajax({
        url: 'https://christoph-lsn.github.io/MT_Site/2-2-1/index.html',
        success: function(data) {
          $('#div id="main-content" class="container goal-2" role="main').html(data);
        }
      });
    });
  </script>
```
But this doesn't work...

My decision_maker.html: https://github.com/christoph-LSN/MT_Site/blob/develop/_pages/decision_maker.html

In general I think I have to  rebuild the indicator page again, but I don't know which files I have to include in my decision:maker.html file.

@sarah-beasley Do you have an idea?

**Comment by brockfanning** [comment]
@christoph-LSN To answer your first question above - yes, all of the pages are "static" in the sense that, after they are built, they are only HTML, CSS, and Javascript.

Regarding showing multiple indicators on a page - I think the challenge there is with the Javascript that handles the chart, table, and map. I suspect that this Javascript assumes there is only one indicator, and so will run into problems when it runs multiple times. That's just a guess - but I'd be really surprised if it worked without problems.

One feature we've discussed in the past would be to allow indicators to be "embedded" as iframes on other pages. With that feature, I think it would be possible for you to embed multiple indicators on a page, without worrying about their javascript colliding with each other.

Technically I think you can already do this, but the problem is that the iframe would contain the header/footer/etc, when you only want the table/chart/map. So I think a feature would need to be added that builds separate versions of each indicator page, in which the extra stuff (header/footer/etc) is hidden.

**Comment by christoph-LSN** [comment]
@brockfanning I did some tests with the iframe. It works more or less. with this Code I get the indicator on my decision_maker page.

```
<iframe id="myIframe" src="https://christoph-lsn.github.io/MT_Site/2-2-1/index.html" width="1024" height="768">
   Sorry your browser does not support inline frames.
</iframe>

<script>
    document.getElementById("myIframe").onload = function() {
        var iframe = document.getElementById("myIframe");
        var iframeDocument = iframe.contentDocument || iframe.contentWindow.document;

        // Header ausblenden
        var header = iframeDocument.querySelector("header[role='banner']");
        if (header) {
            header.style.display = "none";
        }

        // Footer ausblenden
        var disclaimer = iframeDocument.getElementById("disclaimer");
        if (disclaimer) {
            disclaimer.style.display = "none";
        }

        // Navigation ausblenden
        var navigation = iframeDocument.querySelector("nav");
        if (navigation) {
            navigation.style.display = "none";
        }

        // Vorherigen/nächsten Linksteil ausblenden
        var prevNextLinks = iframeDocument.querySelector("previous-next-links");
        if (prevNextLinks) {
            prevNextLinks.style.display = "none";
        }
       
         // Element mit Klasse "heading goal-banner indicator goal-2" ausblenden
        var element = iframeDocument.querySelector(".heading.goal-banner.indicator.goal-2");
        if (element) {
            element.style.display = "none";
        }
        // Element mit der Klasse "footer" ausblenden
        var footer = iframeDocument.querySelector("footer[role='contentinfo']");

        if (footer) {
            footer.style.display = "none";
        }

       var container = iframeDocument.querySelector("container");
    if (container) {
        container.style.display = "none";
       }
    };

</script>
```

I hide header, footer and the goal image. But I can't hide the breadcrumb navigation and the prevous-next-links. Don't understanding whay because it works with the same code I use for hiding the header etc.

![image](https://github.com/open-sdg/open-sdg/assets/94443931/cc4c1a1d-5550-4868-8dad-343c30c8466c)



Two other points are open: I can't using iframe without this vertical scrollbar and when refreshing the site, for a moment, header etc are shown in the iframe.

If I can hide both I think it would be enough for my decision makers. Nevertheless could you tell me, where the indicators are built? so may be I'll try this way too.

**Comment by MoBosse** [comment]
Hi @christoph-LSN ,

I tried to rebuild your page on one of our websites. See the result for two indicators [here](https://dnstestenvironment.github.io/dns-indicators/decision_maker/).

I splitted the code for it into an [html file](https://github.com/DnsTestEnvironment/dns-indicators/blob/develop/_pages/decision_makers.html) and a sepearte [js file](https://github.com/DnsTestEnvironment/dns-indicators/blob/develop/_includes/assets/js/indicatorCleanupForIframe.js) (if you want to do this too you have to include the js file like [this](https://github.com/DnsTestEnvironment/dns-indicators/blob/develop/assets/js/sdg.js#L33)).

Hiding the breadcrumbs can be done like this for example:
`iframeDocument.querySelectorAll(".breadcrumb").forEach(a=>a.style.display = "none");`

Unfortunately, I have not been able to get rid of the slider bars other than setting the height of the iframes to a fixed size. I'd rather make them flexible depending on the height of the content, but for some reason that doesn't seem to be so easy.

**Reply by christoph-LSN** [reply]
@MoBosse Thank you, I will take a look at this. On your site there are noe srollbars in the iframe....

**Comment by christoph-LSN** [comment]
@MoBosse I get rid of the scrollbars wit this code:

 ```
<iframe class="l-no-horizontal-scroll" onload="deleteStuffFromIndicatorPagesForIframe('{{ decision_indicator }}')" id="myIframe_{{ decision_indicator }}" src="https://christoph-lsn.github.io/MT_Site/{{ decision_indicator }}/index.html" width="100%" height="1800px" frameborder="0">
       Sorry your browser does not support inline frames.
```

And I added your JS Code to my script custom.html here: https://github.com/christoph-LSN/MT_Site/blob/develop/_includes/scripts-custom.html so I don't have to override the sdg.js


As you can see here: https://christoph-lsn.github.io/MT_Site/decision_makers/

But I still have the previous-next-links....
inspecting the site this error is shown:

Uncaught TypeError: Cannot read properties of null (reading 'style')
    at deleteStuffFromIndicatorPagesForIframe ((index):522:53)
    at HTMLIFrameElement.onload ((index):244:227)

**Comment by christoph-LSN** [comment]
I also want to hide the metadata tabs, but I don't know which part of the html filr thr meta-tabs are...

**Comment by christoph-LSN** [comment]
@brockfanning I close this, because they (policy decision makers) decided to use Word/Excel
